### ISSUE TITLE
Che 7 GA

### ISSUE THREAD

**Issue by l0rd** [eclipse-che/che#12696]
## Che 7 GA Details

**Schedule**

- 05/22 - Che 7 RC1: Che 7 features completion
- 06/12 - Che 7 RC2: Bug Fixes / Documentation / Test automation
- 07/03 - Che 7 GA (this may change)

**Goal:** Release the new version of Che 7 in a GA state, including documentation, automated tests and website updates. 

**User Story 1 (hosted theia plugin dev)**: As a Che developer I want to start a workspace configured to load theia as the editor and an “hello world” frontend theia plugin (using an existing che stack from the dashboard and with the editor and plugins definition available in a remote plugin registry) so that I will be able to easily developing new plugins.

**User Story 2 (java/nodejs dev)**: As a developer using Che, I want to start a workspace that is configured with the tooling that I defined for my NodeJS/Java project (configuring it from the dashboard, or using a devfile on the repository) so that I will be able to develop on my project and get all language tooling.

**User Story 3 (k8s developer experience)**: As a developer working on a K8S application, I want to be able to use the definition of my application (yaml) in order to create a devfile and get a developer workspace for Che which will replicate my application to allow fast iterative development.

**User Story 4 (chectl and operator)**: 
- As an admin of Che (or user using Che locally), I want to use a CLI based tool that is provided with Che to install and configure my Eclipse Che using (using the operator) so that I can easily control my installation.
- As an admin (or user using Che locally) of a kubernetes cluster, I want to use the OperatorHub to proceed the installation of Eclipse Che.

**User Story 5 (k8s plugin basic flow)**: As a developer using Che, I want to interact with my K8S / OpenShift cluster using dedicated plugins, so I can manage my cluster and push my source code onto the cluster.

**User Story 6 (mutli-editors experience)**: 
- As a developer using Che, I want to build my application using my IDE where I can build code and debug it. I can deploy my application onto K8S / OpenShift cluster using dedicated plugins.
- As a different persona (data scientist), I can work on the same application/context with adapted tools for my needs (Jupyter). 

**Che 7 Success Criteria:**
Implementation of the 5 user stories with:
- The user is able to configure the workspace for his project with the tools he wants
- Complete inner loop is working in the new IDE: Code, Build, Run and Debug
- Devfile has the only format exposed to define workspaces (stacks, factories)
- Languages: Java, Node, Typescript, XML, YAML, C#, Go, Python, PHP
- Out of the box stacks and samples updated
- Chectl as a new CLI for Che
- UX refinements for installation of plugins
- UX adaptations in dashboard to cover devfiles and other che 7 new capabilities
- Ability to switch editors
- Easier monitoring and tracing capabilities
- Pull Request plugin

#### BUGS:
We are using the following label to list all the bugs that need a fix for Che 7 GA.
Link to the GitHub request: [Che 7 GA Bugs](https://github.com/eclipse/che/issues?q=is%3Aopen+is%3Aissue+label%3Atarget%2Fche7GA)

#### UNASSIGNED

- [ ] Self-signed certs support (#12634)

#### IDE2

- [x] **Bug fixing**
  - [x] Setting commands path doesn't work (#11866) 
  - [x] Plugin Preference Schemas are ignored ([#4773](https://github.com/theia-ide/theia/issues/4773)) [PR](https://github.com/theia-ide/theia/pull/4743)
  - [x] Sometimes plugins are not loaded the first time (https://github.com/eclipse/che-theia/issues/101)
  - [x] New Terminal" opens unresponsive terminal window (#13030)
  - [x] Add VS code extensions preferences in workspace configuration (#12992)
  - [x] Rename container view (#12984)
  - [x] Handle plugins with no activate function [#4850](https://github.com/theia-ide/theia/issues/4850)
- [x] Dashboard: remove Plugins/Editors from "create workspace" panel (#12896)
- [x] Plugin installation UX (#12905)
- [x] Sharing from the IDE (#12722)
- [x] Detect services that are listening to a port (#12061)
- [x] Dashboard: cleanup "edit workspace" panel (#12909)
- [x] Dashboard respect Devfile:
  - [x] [UD] Workspace creation from devfile https://github.com/eclipse/che/issues/13058 **[PR](https://github.com/eclipse/che/pull/13469)**
  - [x] [UD] Allow user to manage workspace devfile https://github.com/eclipse/che/issues/13067 
- [x] Make Theia respect Devfile field of Workspace https://github.com/eclipse/che/issues/13112
- [x] Adapt WorkspaceLoader to workspaces created with Devfile https://github.com/eclipse/che/issues/13201
- [x] Dashboard: mark with an error old Che 6 workspaces (#12917)
- [x] Handle Che 7 incompatible factories (#12918)
- [x] Handle memory configuration (#11801) => blocked by OSIO's #12908
- [x] Dashboard, don't display the list of sample plugins by default (#12910) => should be handled by OSIO's [#55](https://github.com/eclipse/che-plugin-registry/issues/55)

#### IDE

- [ ] **Bug fixing**:
   - [ ] Can not run Configured task https://github.com/theia-ide/theia/issues/5064
   - [ ] Recently used tasks section should respect configured tasks https://github.com/theia-ide/theia/issues/5067
   - [ ] Contributed Task Configuration question https://github.com/theia-ide/theia/issues/4928
   - [ ] Provide curated list of containers to run tasks #13390
   - [x] Notifications on Tasks started/ended are missing context #13394
  -  [x] Error running a task/che-commands (https://github.com/eclipse/che-theia/issues/155) **[NEW]**
- [x] wsmaster scalability issues: Output Endpoint (#11961) 
- [x] Che commands should be first class citizens in Theia (#12709):
  - [x] Export commands as task.json (#12712) **[WIP]** [PR](https://github.com/theia-ide/theia/pull/4714)
  - [x] Export debug commands as launch.json (#12710 and #12711) **[WiP]** => [Design document](https://docs.google.com/document/d/1jwbfqbWqTYwE7M0XU51ZQPKS32LASNbPJizFv3Pni_s/edit)
- [x] Finalize Kubernetes Plugin Support (#10574) **[WIP]**  => Already in Che plugin registry [Status document] (https://docs.google.com/spreadsheets/d/1TflFxsQyhySONi63UrHYLbACwVxo8Upq9zOCiFatHG4/edit#gid=667279502) we consider this task like done
- [ ] Finalize OpenShift Plugin Support (#11488) **[WIP]**=> Already in Che plugin registry [Status document](https://docs.google.com/spreadsheets/d/1TflFxsQyhySONi63UrHYLbACwVxo8Upq9zOCiFatHG4/edit#gid=667279502)
- [x] Rework Task extension to a plug-in (#11458) **[WIP]** [merged PR](https://github.com/eclipse/che-theia/pull/97)
- [ ] Integration of the GitHub Plugin (#11867) **[WIP]**
- [ ] Extend pull request flow to add Factory in the comments (#10282)

#### Languages

- [ ] **Bug fixing**:
  - [x] Filesystem privileges issues when bootstrapping JDT LS (#12904)
- [x] VSCode Java Debug support (#12446 and #12395)
- [ ] Test Language Support Plug-ins on Minishift (#12941)
- [x] Python support:
  - [x] Integration of Python (#12771)
  - [x] Python Stack (#12944)
  - [x] Django Stack (#12945) 
- [x] C# support:
  - [x] Integration of C# - https://github.com/eclipse/che/issues/12773
  - [x] C# Stack #13096
- [ ] PHP support:
  - [ ] Integration of PHP (#12772)
  - [ ] PHP Stack (#12620)
- [x] Camel support:
  - [x] Integration of Camel Plugin (#12584)

#### OSIO
- [x] Enforce a better plugin registry naming convention https://github.com/eclipse/che-plugin-registry/issues/55 **[Done in 7 beta 4]**
   - [x] che-incubator/chectl#89
   - [x] redhat-developer/rh-che#1390
- [x] Merge meta.yml and che-plugin.yml - #12908 **[Done in 7 beta 5]**
   - [x] https://github.com/eclipse/che-plugin-registry/pull/135
   - [x] https://github.com/redhat-developer/rh-che/issues/1308
   - [x] Update version of registry and plugin broker on wsmaster side to catch up regitry / broker changes - https://github.com/eclipse/che/pull/13311
- [x] VS Code marketplace rate limit - #12840 **[WORKAROUND PROVIDED]** 
- [x] Allow to use 'latest' for plugin versions #12937 **Sprint: 166**
   - [x]  redhat-developer/rh-che#1395
   - [ ] **NOT PLANNED FOR GA** redhat-developer/rh-che#1396 PR sent https://github.com/eclipse/che/pull/13414 **but conflicts with** https://github.com/eclipse/che/pull/13297 
- [x] Support running multiple VS Code extensions in the same sidecar #12395
- [x] Avoid the need to set userland-proxy #11018 **Sprint: 166**
- [x]  Implement compatibility with the new automatic multicast discovery in plugin broker https://github.com/redhat-developer/rh-che/issues/1354 **Sprint: 166** 
- [ ] ~~Telemetry - https://github.com/eclipse/che/issues/12717~~ **Postponed after Che 7 GA**
- [ ] **Bug fixing**:
  - [ ] Use upstream che 7 stacks as-is in rhche without any customizations https://github.com/redhat-developer/rh-che/issues/1347
  - [ ] Most public images do not work as the base for Che 7 workspaces on OpenShift #13454 
  - [ ] Need to make sure that all devfiles from registry work on che.openshift.io https://github.com/redhat-developer/rh-che/issues/1434
  - [x] '?' folder is created when building console-java-simple project using java-maven Che 7 stack https://github.com/eclipse/che/issues/13451
  - [ ] Workspace failed to start because of plugins installation process failed #12999 **NO STEPS TO REPRO**

#### Platform

- [x] Add more fields to the plug-in registry - https://github.com/eclipse/che-plugin-registry/issues/71
- [ ] Finalize Devfile - https://github.com/eclipse/che/issues/11549 
  - [x] Provide database layer storage and integrity of devfile as part of workspace | https://github.com/eclipse/che/issues/13075 **Sprint: 165**
  - [x] Add an ability to use Devfile as workspace configuration as alternative to WorkspaceConfig | https://github.com/eclipse/che/issues/13079 **Sprint: 165**
  - [x] Perform schema and integrity validation on creating a workspace from Devfile https://github.com/eclipse/che/issues/13315 **Sprint: 167**
  - [x] [POC] tasks and launch configuration as part of Devfile | https://github.com/eclipse/che/issues/13057   **Sprint: 166**
  - [x] Set memory limit for sidecar plugins on devfile | https://github.com/eclipse/che/issues/12981  **Sprint: 165**
  - [x] Introduce registryUrl field for cheEditor/chePlugin components | https://github.com/eclipse/che/issues/13104 **Sprint: 167**
  - [x] Rename `name` field of component to Devfile to `alias` | https://github.com/eclipse/che/issues/13103 **Sprint: 165**
  - [ ] Adopt devfile format for Stacks - https://github.com/eclipse/che/issues/12924  **Sprint: 167**
     - [x] Initial setup of Che Defvile registry | https://github.com/eclipse/che/issues/13353 **Sprint: 167**
     - [ ] StackService decommissioning  https://github.com/eclipse/che/issues/13459 **Sprint: 167**
     - [x] Che 7 stacks should use version "latest" of plugins https://github.com/eclipse/che/issues/13426
     - [x] Deploy and use public che devfile registry on https://che.openshift.io infrastructure  https://github.com/eclipse/che/issues/13355 **Sprint: 167**
  - [x] Adopt devfile format for workspaces(API part) - https://github.com/eclipse/che/issues/12926 **Sprint: 165**
  - [x] [POC] Devfile format for Stack, Factory, Workspace API | https://github.com/eclipse/che/issues/13018 **Sprint: 165**
  - [x]  Provide database layer storage and integrity of devfile as part of workspace | https://github.com/eclipse/che/issues/13075 **Sprint: 165**
  - [x] Make Devfile syntax K8s-compliant https://github.com/eclipse/che/issues/13061
     - [x] [Devfile] Make specVersion and name more K8s-like https://github.com/eclipse/che/issues/13336  **Sprint: 167**
- [x] Finalize Monitoring - https://github.com/eclipse/che/issues/10329
  - [x] Create documentation for monitoring use-case | https://github.com/eclipse/che/issues/12543  **Sprint: 166**
- [x] Finalize Tracing - https://github.com/eclipse/che/issues/10298
  - [x] Document tracing usecase  https://github.com/eclipse/che/issues/12545 **Sprint: 167**
- [x] Support mountSources for non-plugin tools in devfile https://github.com/eclipse/che/issues/12554 **Sprint: 168**
- [ ] Deprecate storing of workspaces converted from Devfile https://github.com/eclipse/che/issues/13423  **Sprint: 166**
- [ ]  Legacy environment variables cleanup https://github.com/eclipse/che/issues/12679  **Sprint: 167**
- [ ] Bug fixing:
  - [x] Error when starting workspace, created from devfile | https://github.com/eclipse/che/issues/13040  **Sprint: 165**
  - [x] NPE after force infrastructure restart https://github.com/eclipse/che/issues/13313 **Sprint: 165**
  - [x] GitHub oAuth authorization does not work with Red Hat SSO 7.3 https://github.com/eclipse/che/issues/13380  **Sprint: 167**
  - [x] Workspace is not starting, when k8s StorageClass has volumeBindingMode=WaitForFirstConsumer https://github.com/eclipse/che/issues/12889 **Sprint: 167**
  - [x] Volume created but the secret with the self-signed cert is not https://github.com/eclipse/che/issues/13433
#### Installation

- [x] Operator for Che 7 on OperatorHub (#11469) - **[WIP]** [PR1](https://github.com/operator-framework/community-operators/pull/383) and [PR2](https://github.com/operator-framework/community-operators/pull/383)
- [ ] chectl:
  - [x] command `server:stop`
  - [x] command `server:delete`
  - [x] command `devfile:generate` ([#81](https://github.com/che-incubator/chectl/issues/81)) - **[WIP]** [PR](https://github.com/che-incubator/chectl/pull/84)
  - [x] OCP support ([#80](https://github.com/che-incubator/chectl/issues/80))
  - [x] Docker desktop support ([#14](https://github.com/che-incubator/chectl/issues/14))
  - [ ] command `server:update` ([#32](https://github.com/che-incubator/chectl/issues/32) and #12411))
  - [ ] Bug fixing: [#18](https://github.com/che-incubator/chectl/issues/18), [#82](https://github.com/che-incubator/chectl/issues/82), [#78](https://github.com/che-incubator/chectl/issues/78)

#### QE

- [x] Define the strategy to add tests on Theia (WIP) (Investigate e2e technologies for Thea #12204)
- [ ] E2E tests suite for Che 7 (#12728) - **[WIP]**
  - [x] Setup the new test solution (selenium-typescript + Mocha + Chai) (#13114) - **[WIP]**  - PR #13200
  - [x] Prepare yaml file for deploying test application (Nodejs+MongoDb) based on Kubernetes/Openshift infra (#13119) - **[WIP]**
  - [x] Add PR trigger for happy path job (`che7-test-ci`) (#13173) - **[WIP]** 
  - [x] Setup periodic job to test Che started with chectl (#13229) - **[WIP]** 
  - [x] Configure CI for new E2E tests (#12729)
- [ ] Scalability testing - **[WIP]**
- [ ] Add theia-ide/theia PR check against che.osio (#13258)
- [ ] Add an eclipse/che nightly test against che.osio (#13259)
- [ ] Add eclipse/che PR check test against che.osio (#13260)
- [ ] Run Happy path test as PR check on kubernetes infra (#13473)

#### Documentation, training and demos

- [ ] Integrate new documentation about extensibility 
- [ ] Port Che6 documentation to Che 7
- [ ] Update README
- [ ] Update Contributing wiki
- [ ] Release notes for all new features
- [ ] Make some demo/workshop material (doc + short blog post + short video) [github repo](https://github.com/che-incubator/che-demos)
  - [x] Devfile (@sleshchenko) https://github.com/che-incubator/che-demos/pull/3
  - [ ] New multi container based workspace model
  - [ ] VS Code plugin compatibility
  - [ ] Container based development (@eivantsov)
  - [ ] chectl
  - [ ] Che operator
  - [x] Monitoring and Tracing (@ibuziuk)
  - [ ] Editor swapping capability (@skabashnyuk)

#### Marketing and Promotion

- [ ] Update Eclipse Che website
   - [ ] Put links for Che 7 download / signup on the homepage
   - [ ] Create videos to demonstrate major Che 7 features (need a list) 
   - [ ] Update Technology page and combine Features
   - [ ] Update Extend page
   - [ ] Update Technology page
   - [ ] Update footer links (remove Red Hat link)
   - [ ] Remove Community page and redirect URL to https://github.com/eclipse/che/blob/master/CONTRIBUTING.md
- [ ] Blog post series of major Che 7 features (need a list) 
- [ ] Dave: Collect testimonials for Che 7 (user, OEM, plug-in developer)
- [ ] "Getting started with Che 7" document (part of project docs? or elsewhere?)
- [ ] Jesse: Create sample tweets to use around Che 7 release for #chetips
- [ ] Stevan/Dave: Prepare webinar to present Che 7

PR/AR
- [ ] Draft release announcement for a press release
- [ ] Prepare analyst briefing for Che 7 (if AR agrees it is a good idea)

**Comment by gazarenkov** [comment]
Kamino cloned this issue to [gazarenkov/che-ide6](https://github.com/gazarenkov/che-ide6/issues/22)

**Comment by tsmaeder** [comment]
Discovered https://github.com/theia-ide/theia/issues/4850 which is a possible blocker for php-debug support.

**Comment by l0rd** [comment]
@tsmaeder added in the IDE2 backlog

**Comment by l0rd** [comment]
This issue has been replaced by the endgame document https://github.com/eclipse/che/issues/13637
